PURL-TYPES.rst

```diff
@@ -221,6 +207,34 @@ nuget
 
       pkg:nuget/EnterpriseLibrary.Common@6.0.1304
 
+oci-artifact
+------------
+``oci-artifact`` for all artifacts stored in OCI Distribution-spec conformant
+registries including container images built by docker and others:
+
+- There is no canonical package repository for OCI artifacts. Therefore
+  ``oci-artifact`` pURLs must be registry agnostic.
+- OCI pURLs do not require a ``namespace``. The ``namespace`` is specific
+  to the physical location of the package.
+- The ``name`` is not case sensitive and must be lowercased. The name is the
+  last fragment of the repository name. For example if the repository
+  name is ``library/debian`` then the ``name`` is ``debian``.
+- The ``version`` is the ``@sha256:digest`` of the artifact and is strongly
+  suggested for oci artifacts.
+- Optional qualifiers may include:
+
+  - ``arch``: key for a package architecture, when relevant
+  - ``repository``: A repository URL where the artifact may be found, but not intended as the only location
+  - ``tag``: artifact tag that may have been associated with the digest at the time
+  - ``type``: type of artifact stored in an OCI registry
+- ``subpath`` is not relevant to ``oci-artifact``
+- Examples::
+
+      pkg:oci-artifact/debian@256:<shasum>?repository=docker.io/library/debian&arch=amd64&type=application/vnd.oci.image.manifest.v1+json
+      pkg:oci-artifact/debian@sha256:<shasum>?tag=bullseye&type=application/vnd.oci.image.manifest.v1+json
+      pkg:oci-artifact/metallb@sha256:<shasum>?repository=registry.io/bitnami&type=application/vnd.cncf.helm.config.v1+json
+      pkg:oci-artifact/hello-wasm@256:<shasum>?type=application/vnd.wasm.config.v1+json&tag=v1
```

Comment:
Per the PURL spec, PURL is used to "reliably identify and locate software"

I don't see how anyone can locate using this purl.

Reply:
Is the `repository` URL identifier not enough?

Reply:
This is a valid concern, and also something I've discussed with @iamwillbar several times: In cloud native, there isn't a single ecosystem akin to other package managers. For example, there is no "docker" ecosystem similar to a "pip" ecosystem. Container registries are distributed, there are many client tools, and the artifacts can be any number of things. The only thing that ties them together is the OCI specification. Hence the "oci-artifact" type.
We would also like to not link the pURL to a specific domain or tool as they can move from cloud to cloud and still be the same artifact.

Reply:
I understand the distributed nature of oci artifacts. My point is with line 236, which does not have a `repository` specified and oci-artifact does not have a default. Therefore it is impossible to locate the example specified on line 236

```pkg:oci-artifact/hello-wasm@256:<shasum>?type=application/vnd.wasm.config.v1+json&tag=v1```

IMO, repository should be mandatory for oci-artifact purl types. Without it, they're not formatted in the spirit of purl which is to "reliably identify and locate software"

Reply:
@stevespringett do you think it could be "strongly recommended" that one location to retrieve the artifact is provided in the `repository_url`. There could be scenarios where you want to share the identity but not a location (e.g. the location is currently an internal/private location or not yet known, the advantage of that is that adding a `repository_url` later doesn't invalidate the identity portion of the purl.

Reply:
If private repos are used, they should be specified. Without a repository_url (or a default) it will be impossible to locate the artifact, so organizations will not be able to perform any kind of evaluation of the artifact (outdated, tampered, verification, etc).  

This ticket was framed in the context of SBOM, and security is the only reason why we're discussing SBOM. Therefore, if a new purl type is supported without the ability to locate the artifact, we won't be able to support security use cases. This doesn't seem like a good strategic decision. 

To my understanding, the only reason why oci-artifact is necessary is to support artifacts other than container images. Everything else can easily be supported with the existing Docker type with minor changes.  IMO, we may want to position oci-artifact as a superset of Docker which means oci-artifact should inherit Dockers defaults, including repository_url. If a NULL repository_url is truly wanted, then I think we need to define an optional way to specify that which would be applicable to every purl type, but it should not be the default simply by omitting repository_url as is the current case with this PR.

Reply:
> the only reason why oci-artifact is necessary is to support artifacts other than container images

It's true that we need oci-artifacts to support artifacts other than container images but also, describing all container images as `docker` is not correct. Docker is not a container image format, it's just the build tool.  As @nishakm mentioned above, even Docker uses the OCI image specification as the format and we should properly refer to this in purl.

> If a NULL repository_url is truly wanted, then I think we need to define an optional way to specify that which would be applicable to every purl type

No objection on this -- let's make this possible! A NULL repository has use cases outside of just oci artifacts as private repositories exist for other package types as well.

Reply:
I'm hesitant for other ecosystems where the identity isn't a content address like it is for `oci-artifact`. If I take a package identity to two different registries I could get different packages back for some ecosystems, so I'd recommend only allowing  `repository_url` to be excluded if the package manager's purl has some way of guaranteeing it's the same package.

A NULL repository for `oci-artifact` may make sense when building the SBOM before you've published to a registry and don't know the "final" destination for the image. As I mentioned above a `repository_url` can be added without breaking the package identity, so updated SBOMs could be issued downstream when a specific canonical registry is known.

Reply:
> I'm hesitant for other ecosystems where the identity isn't a content address like it is for `oci-artifact`. If I take a package identity to two different registries I could get different packages back for some ecosystems, so I'd recommend only allowing `repository_url` to be excluded if the package manager's purl has some way of guaranteeing it's the same package.

Fair enough. To me, the "strongly recommended" approach makes the most sense but want to address all concerns. Cloud native artifacts do not compare "apples to apples" to typical package types with package managers so a slight deviation from the "spirit of purl" for this feels reasonable. The proposal for oci-artifact still ensures that the software can be "reliably identified".

> A NULL repository for `oci-artifact` may make sense when building the SBOM before you've published to a registry and don't know the "final" destination for the image. 

Or when the repository is private and you're unable to include it for security reasons, in which case, it may remain absent. But, as you mentioned, this would not affect the package identity.

Reply:
OCI artifacts i.e. kinds of artifacts defined by OCI

Reply:
In that case, I'd rename this type to `oci` to avoid confusion with OCI Artifacts.

Reply:
Would the oci type in this case only be used to describe only oci container images?

Reply:
@jonjohnsonjr, this was discussed in the last oci meeting, and you agreed to use oci-artifact. OCI was too general. the oci container image type is logically one of the OCI artifact types.